mantidproject/mantid
# Reflectometry GUI restructuring 4: Runs tab

**Description of work.**
Ongoing work to integrate the `JobTreeView` into the `ISIS Reflectometry` GUI. This PR is being merged to a feature branch because not all functionality is implemented yet.

This PR restructures the Runs Tab view/presenter for the GUI to follow the same structure as the other tabs which have already been restructured as part of #23630 and #24345.

- Rename the `...ReflRunsTab...` components to just `Runs...`
- Move the components to the `GUI/Runs/` subdirectory
- Add a new `RunsViewSubscriber` interface and use this inside the view rather than the presenter.
- Add methods to `RunsTablePresenter` and underlying classes to support removing all rows
- Change to `JobTreeView` to allow deleting the last selected row - it now leaves an empty row instead of throwing.

Note that MVP models have not been added at this point but this will be done in a later PR.

**Report to:** Nobody

**To test:**

Note that this is work in progress on a feature branch so not all functionality is available. Here are some suggestions for tests:
- Open the `ISIS Reflectometry` interface.
- Use the buttons to add some rows in the table and then delete them - check that it no longer crashes
- Try entering an investigation ID e.g. `INTER` and `1120015` and click `Search`. Note that catalog login is not currently implemented so you need to first login to the catalog via the main Catalog menu.
- The search should succeed and show all the runs for the investigation. Select some of them and click `Transfer`. They should be added to the main table and grouped by title with runs for different angles on different rows. If runs have the same title and angle they will appear on the same line as summed, e.g. `11935+11936`
- Click `Autoprocess`. This will copy all runs for the experiment into the main table and start an endless progress bar. Otherwise it is not implemented and currently does nothing. Note that Pause is not currently implemented.

** Known issues **
- If you copy a group then paste it onto a child row of an existing group, the tree extends to more than 2 levels deep (it should be 2 max)
- Processing and Autoprocess are not implemented
- Catalog login is not implemented

Refs #23027

*This does not require release notes* because it only contains under-the-hood changes

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
